### ISSUE TITLE
Question about logit_bias/mask

### ISSUE THREAD

**Issue by zhangyuhanjc** [eth-sri/lmql#258]
## Question
I have noticed that during the use of LMQL, the client-side often sends a large number of logit_bias, even though there are no relevant constraints in my WHERE statement. These logit_bias values can sometimes affect the reasoning results of the model. I would like to know why these unexpected logit_bias values occur and how to avoid them
## My query 
```
    argmax(chunksize=1)
        """A:what's the weather like today? B:[resp]"""
    from
        lmql.model("...model_path" ,cuda=True,load_in_4bit=True,trust_remote_code=True,endpoint="localhost:8081")
    where
        len(resp)<30
```
## logit_bias each step i print
The actual inference process involves several dozen steps, and I have extracted and printed the logit_bias of a few intermediate steps

interpreter.where_for_sequence logit_mask (array([  462,  1669,  2683,  4623,  5106,  5285,  5314,  6121,  6336,
        6590,  6675,  6688,  6917,  7119,  7313,  7775,  8004,  8954,
        9166,  9863, 10760, 11511, 12084, 12746, 12965, 13383, 13852,
       14209, 14235, 15125, 15428, 15746, 15909, 16068, 16167, 16592,
       16654, 16746, 16845, 16951, 17443, 17489, 17517, 17865, 17994,
       17996, 18189, 18383, 18822, 18884, 18995, 19185, 19220, 19270,
       19285, 19511, 19838, 19854, 20110, 20240, 20360, 20528, 20660,
       20706, 20837, 20999, 21088, 21097, 21382, 21666, 21699, 21702,
       21862, 21992, 21995, 22489, 22522, 22540, 22816, 22893, 22920,
       23055, 23134, 23236, 23303, 23409, 23491, 23949, 24118, 24225,
       24305, 24496, 24521, 24637, 24771, 25012, 25176, 25285, 25499,
       25700, 25844, 25913, 26227, 26278, 26295, 26358, 26374, 26588,
       26702, 26845, 26916, 27097, 27241, 27443, 27577, 28163, 28244,
       28574, 28650, 28733, 28757, 29109, 29303, 29304, 29380, 29577,
       29605, 29643, 29694, 29769]),) 
interpreter.where_for_sequence logit_mask (array([  462,   632,   795, ..., 29839, 29864, 29868]),) 
interpreter.where_for_sequence logit_mask (array([  462,   632,   795, ..., 29839, 29864, 29868]),) 
interpreter.where_for_sequence logit_mask (array([  462,   632,   795, ..., 29839, 29864, 29868]),) 
interpreter.where_for_sequence logit_mask (array([  462,   632,   795, ..., 29839, 29864, 29868]),) 
interpreter.where_for_sequence logit_mask (array([  462,   632,   795, ..., 29839, 29864, 29868]),) 

## Decode those mask
i try to decode those masks, however, they appear to be difficult to comprehend.
```
tokenizer.decode(torch.tensor([462,  1669,   ......, 29694, 29769]))
```
when i decode thos logit_bias/mask, i got
'                             ---------------- International documentation configuration implementation international automatically corresponding EinzelnachweisespringframeworkViewControllergetElementByIdincludegraphics**************** understanding representation================ functionality authentication Unfortunately communicationAndroidRuntime classification################ transformation circumstances approximately Championships unfortunatelyActivityThreadOnClickListenereltemperaturen MassachusettsaddEventListenerprintStackTraceAuthentication approximation significantly characteristic Internacional administration initialization Bibliographieamerikanischer Entertainmentdisambiguation distributions                estadounidense administrative consideration universitaire internacional//////////////// Unterscheidung interpretation internationale implementations Alternatively autorytatywna distinguished contributions straightforward Configurationadratkilometer representativeimplementation multiplication simultaneously relationships characteristics specification principalmentepreventDefault investigation representations partiellementLongrightarrow configurationsauthentication responsibility Classification Administration административ Fußballspieler contradiction Schriftstellerlongrightarrow développement possibilities compatibility Transfermarkt collaboration establishment verschiedenen................ independently organizations StringBuilder notifications decomposition modifications veröffentlichtLayoutInflaterRelativeLayoutCollectionView concentration computational setContentView________________ Authentication administrator participation extraordinary französischen Mediabestanden Genomsnittlig authorization RepresentativesPropertyChanged recommendation transformations environmental stackoverflowbackgroundColor neighbourhood interpolation identification'

**Comment by lbeurerkellner** [comment]
As far as I can see, your constraint `len(resp) < 30` indeed does require logit biasing, as it operates on a character level.

It may be useful to point out that `len(resp) < 30` [constrains the model on a character level](https://lmql.ai/docs/language/constraints.html#character-length), i.e. it means the response should not exceed more than 30 characters. However, depending on the current decoding offset, some (very long) tokens may already exceed the `30` character limit. With every added token, this becomes more severe, as you will have generated some character already, leading to a further reduction of allowed continuation tokens, as to not exceed the overall limit of 30 characters. Eventually, you may even end up with a very strict mask, that only allows single-character tokens as final continuation.

If you just want to constrain the number of tokens, use `len(TOKENS(resp)) < 30`. This will not require logit biasing, as each token just counts as one and is thus also much more efficient.

**Comment by zhangyuhanjc** [comment]
> As far as I can see, your constraint `len(resp) < 30` indeed does require logit biasing, as it operates on a character level.
> 
> It may be useful to point out that `len(resp) < 30` [constrains the model on a character level](https://lmql.ai/docs/language/constraints.html#character-length), i.e. it means the response should not exceed more than 30 characters. However, depending on the current decoding offset, some (very long) tokens may already exceed the `30` character limit. With every added token, this becomes more severe, as you will have generated some character already, leading to a further reduction of allowed continuation tokens, as to not exceed the overall limit of 30 characters. Eventually, you may even end up with a very strict mask, that only allows single-character tokens as final continuation.
> 
> If you just want to constrain the number of tokens, use `len(TOKENS(resp)) < 30`. This will not require logit biasing, as each token just counts as one and is thus also much more efficient.

I understand now, thank you very much for your clear explanation. The logit_bias is back to normal after changing to len(TOKENS(resp)) < 30

**Comment by zhangyuhanjc** [comment]
close
